### PR TITLE
feat: Filter plugins by matching Action

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

See https://github.com/PostHog/posthog/issues/21721
and https://github.com/PostHog/product-internal/issues/599

We want to replace the existing action webhooks with something a little more fit-for-purpose. To do this we need to:
1. Support filtering plugins by actions as a config option on the plugin
2. Allow webhook plugins to take care of templating the event into valid webhook content (e.g. for a slack webhook)

This PR attempts to solve problem 1

## Changes

- [x] Refactors the ActionMatcher into the hub to be used by both webhook firing and composeWebhook processing
- [x] Moves reloading and pubsub into the ActionManager class as it feels more part of it.
- [x] Filters `composeWebhoook` plugins by the relevant action matcher if necessary
- [x] Adds option to the (legacy) UI behind a flag
- [ ] Add tests for filtering logic
- [ ] Ensure actions can not be deleted if they have connected plugins

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: plugin-server/src/main/pluginsServer.ts [hunk 1]
```diff
@@ -361,38 +357,18 @@ export async function startPluginsServer(
...
-                    serverConfig.EXTERNAL_REQUEST_TIMEOUT_MS
-                )
-            const appMetrics =
-                hub?.appMetrics ??
-                new AppMetrics(
-                    KafkaProducerWrapper,
-                    serverConfig.APP_METRICS_FLUSH_FREQUENCY_MS,
-                    serverConfig.APP_METRICS_FLUSH_MAX_QUEUE_SIZE
-                )
+            ;[hub, closeHub] = hub ? [hub, closeHub] : await createHub(serverConfig, capabilities)
```
Comment (tiina303) [inline plugin-server/src/main/pluginsServer.ts:360]: If you check out the blame you can see https://github.com/PostHog/posthog/commit/d77858d275e0cc03ac5ea429eb81334cab7af22e we're wanting to rather move away from hub than add more stuff to it.

File: plugin-server/src/utils/db/hub.ts [hunk 2]
```diff
@@ -151,6 +153,10 @@ export async function createHub(
         serverConfig.EXTERNAL_REQUEST_TIMEOUT_MS
     )
 
+    const actionManager = new ActionManager(postgres, serverConfig)
+    await actionManager.start() // TODO: Do we want this to happen here?
+    const actionMatcher = new ActionMatcher(postgres, actionManager)
```
Comment (tiina303) [inline plugin-server/src/utils/db/hub.ts:158]: we probably don't want this here - this is only relevant to webhooks and export plugins, not all the other plugin-server instances

File: plugin-server/src/worker/plugins/run.ts [hunk 3]
```diff
@@ -353,15 +353,32 @@ export async function runPluginTask(
...
+                // TODO: Add matchActions option
+                if (pluginConfig.matchActions.length === 0) {
+                    const matchedActions = await hub.actionMatcher.match(event)
+                    return pluginConfig.matchActions.some((actionId) => matchedActions.find((x) => x.id === actionId))
+                }
+
+                return pluginConfig
+            })
+        )
+    ).filter(Boolean) as PluginConfig[]
```
Comment (tiina303) [inline plugin-server/src/worker/plugins/run.ts:377]: I'd only add this capability to `composeWebhook` plugins and I'd suggest we do this within `runSingleTeamPluginComposeWebhook` <- would make fixing tests a bunch easier probably too

**Review by tiina303 — COMMENTED** [review]
Thoughts on the plan:
😬  Refactors the ActionMatcher into the hub to be used by both webhook firing and processing
👍  Moves reloading and pubsub into the ActionManager class as it feels more part of it.
😬  "Starting" the action manager happens in the hub creation - should it?

> Currently actions are only loaded if they have a configured webhook. I guess we should change it so that they are loaded if an enabled plugin is referencing it?

I'm not sure if this would be better or worse, but we could create a `composeActionWebhook` plugin function or whatever we name it, that only has actions list as a trigger option, it being a separate function could make it easier to distinguish it in the UI too 🤷‍♀️  Then we can later migrate all the other plugins over to it if that's the future.

👍  Actually do the filtering of plugins to be called
👍  Fix all the tests

> Understand the impact of this additional filtering logic (looks like if there is a cohort involved it hits postgres every time?)

We might want to grandfather existing webhooks, but not allow any new usage of stuff that would be very expensive 🤷‍♀️  We can hide it from users initially and test it out for team 2 first, then make calls about this part.

👍  Add the UI to support filtering by action as a plugin config option

File: plugin-server/src/main/pluginsServer.ts (hunk 1 above)
Comment (benjackwhite) [reply plugin-server/src/main/pluginsServer.ts:360]: Hmmm. Not meaning to get into a big refactor but the "hub" is essentially your dependency injection. Getting around it seems to be making the code pretty hard to parse and manage 😅 Do you have an alternative solution? Or is it just the nasty passing of variables everywhere?

File: plugin-server/src/worker/plugins/run.ts (hunk 3 above)
Comment (benjackwhite) [reply plugin-server/src/worker/plugins/run.ts:377]: Ah shame. Feels like a core functionality but fair enough.

Comment (tiina303) [reply plugin-server/src/worker/plugins/run.ts:377]: there are only two supported kinds of plugins: 
- processEvent  - doesn't make sense to have any filtering here by actions as the actions might not match at that point the same way (the event doesn't have geoIP info yet for example before running geoIP).
- composeWebhook - where we'll be adding this.

We still have some onEvent apps too that we haven't migrated, but there won't be new ones.

File: plugin-server/src/main/pluginsServer.ts (hunk 1 above)
Comment (tiina303) [reply plugin-server/src/main/pluginsServer.ts:360]: looking at this again - not sure I follow on the big refactor part, we'll want to not initialise the action matcher in the hub, so moving it here and starting it here would be easy, no?

Comment (tiina303) [reply plugin-server/src/main/pluginsServer.ts:360]: so in this code path we're just creating a hub to throw it away immediately? 
Instead, I'd move this here after PG creation, or am I missing something?
```
    // we need to create them. We only initialize the ones we need.
    const postgres = hub?.postgres ?? new PostgresRouter(serverConfig)
    const actionManager = new ActionManager(postgres)
    await actionManager.prepare()
    const actionMatcher = new ActionMatcher(postgres, actionManager)
```

File: plugin-server/src/utils/db/hub.ts (hunk 2 above)
Comment (tiina303) [reply plugin-server/src/utils/db/hub.ts:158]: maybe we initialize these to undefined and init them in `startAsyncWebhooksHandlerConsumer` and `startAsyncOnEventHandlerConsumer`?

File: plugin-server/src/main/pluginsServer.ts (hunk 1 above)
Comment (benjackwhite) [reply plugin-server/src/main/pluginsServer.ts:360]: So in that area the code makes no sense to me at all 😅 The `hub` is basically never there and if it isn't we have all the duplocate setup code. For reasons that are super unclear as the hub is mostly just instantiating things (like a DI)

Comment (tiina303) [reply plugin-server/src/main/pluginsServer.ts:360]: If the plugin-server being set-up has ingestion capabilities (e.g. local dev, hobby deploy), then hub is created in an earlier if statement. 

We want to move away from the hub and have more granular dependency injection, this is what Harry's PR was in direction for. We're approaching this via "leave the campsite cleaner than you found it".

File: plugin-server/src/types.ts [hunk 4]
```diff
@@ -666,17 +671,6 @@ export interface ClickHouseEvent extends BaseEvent {
 /** Event in a database-agnostic shape, AKA an ingestion event.
  * This is what should be passed around most of the time in the plugin server.
  */
```
Comment (tiina303) [inline plugin-server/src/types.ts:673]: 👍  like what you did here ... less types is better :) And the `PreIngestionEvent` indeed matches to `PostHogEvent` almost, extended the comment, so they don't diverge too much in the future.
```suggestion
 * Keep easy to convert to/from PostHogEvent (in plugin-scaffold repo) that's used in plugins.
 */
```

File: posthog/models/plugin.py [hunk 5]
```diff
@@ -252,8 +252,14 @@ class Meta:
     # Used in the frontend
     name: models.CharField = models.CharField(max_length=400, null=True, blank=True)
     description: models.CharField = models.CharField(max_length=1000, null=True, blank=True)
-    # Used in the frontend to hide pluginConfgis that user deleted
+    # Used in the frontend to hide pluginConfigs that user deleted
     deleted: models.BooleanField = models.BooleanField(default=False, null=True)
+    match_action = models.ForeignKey(
```
Comment (tiina303) [inline posthog/models/plugin.py:257]: could you add a comment above this what this field is/does

optional: we could have multiple actions here via [ManyToMany](https://docs.djangoproject.com/en/5.0/topics/db/examples/many_to_many/), though we'd need to update https://github.com/PostHog/posthog/blob/master/plugin-server/src/utils/db/sql.ts#L94C23-L94C42 & `runComposeWebhook` accordingly too. We can initially still limit it to be a single action in the UI. I'm not sure if you're thinking longer term you'll want multiple actions/events to be able to be picked or not 🤷‍♀️  it's not that different here at this point to make it work over multiple.

File: plugin-server/src/worker/plugins/run.ts [hunk 6]
```diff
@@ -210,7 +228,7 @@ export async function runComposeWebhook(hub: Hub, event: PostHogEvent): Promise<
 
 export async function runProcessEvent(hub: Hub, event: PluginEvent): Promise<PluginEvent | null> {
     const teamId = event.team_id
-    const pluginMethodsToRun = await getPluginMethodsForTeam(hub, teamId, 'processEvent')
+    const pluginMethodsToRun = await getPluginMethodsForTeam(hub, event.team_id, 'processEvent')
```
Comment (tiina303) [inline plugin-server/src/worker/plugins/run.ts:231]: nit
```suggestion
    const pluginMethodsToRun = await getPluginMethodsForTeam(hub, teamId, 'processEvent')
```

**Review by tiina303 — COMMENTED** [review]
We discussed plugin-server/src/worker/plugins/run.ts in a huddle, checked out other changes 👍

File: posthog/models/plugin.py (hunk 5 above)
Comment (benjackwhite) [reply posthog/models/plugin.py:258]: I don' think many to many will make much sense tbh. Makes it simpler this way, and you can solve many to many by multiple match groups in the action.

File: plugin-server/src/utils/event.ts [hunk 7]
```diff
@@ -181,18 +182,3 @@ export function formPipelineEvent(message: Message): PipelineEvent {
     })
     return event
 }
-
-export function formPluginEvent(event: RawClickHouseEvent): PluginEvent {
```
Comment (benjackwhite) [inline plugin-server/src/utils/event.ts:185]: This wasn't referenced anywhere

File: plugin-server/src/worker/plugins/run.ts [hunk 8]
```diff
@@ -351,17 +356,59 @@ export async function runPluginTask(
...
     }
+
     const methodsObtained = await Promise.all(
         pluginConfigs.map(async (pluginConfig) => [pluginConfig, await pluginConfig?.vm?.getVmMethod(method)])
     )
-    return methodsObtained as [PluginConfig, VMMethods[M]][]
+
+    const methodsObtainedFiltered = methodsObtained.filter(([_, method]) => !!method) as [
+        PluginConfig,
+        VMMethodsConcrete[M]
```
Comment (benjackwhite) [inline plugin-server/src/worker/plugins/run.ts:375]: This is a nice improvement to skip the duplicate filtering and simplify a lot of the other logic by making the selected method "concrete".

**Comment by benjackwhite** [comment]
@tiina303 I get a lot of failing tests due to connections being closed and I don't know why. Any idea what might be happening?
